File: app/src/main/java/com/rokudo/xpense/data/retrofit/GetDataService.java
package com.rokudo.xpense.data.retrofit;

import com.rokudo.xpense.data.retrofit.models.AccountDetails;
import com.rokudo.xpense.data.retrofit.models.EndUserAgreement;
import com.rokudo.xpense.data.retrofit.models.Institution;
import com.rokudo.xpense.data.retrofit.models.Requisition;
import com.rokudo.xpense.data.retrofit.models.Token;
import com.rokudo.xpense.data.retrofit.models.Transactions;
import com.rokudo.xpense.data.retrofit.models.TransactionsResponse;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface GetDataService {

    @GET("institutions/?country=RO")
    Call<List<Institution>> getAllInstitutions();

    @FormUrlEncoded
    @POST("token/new/")
    Call<Token> getToken(@Field("secret_id") String secret_id, @Field("secret_key") String secret_key);

    @FormUrlEncoded
    @POST("agreements/enduser/")
    Call<EndUserAgreement> createEUA(@Field("institution_id") String institution_id,
                                     @Field("access_scope") List<String> access_scope);


    @FormUrlEncoded
    @POST("requisitions/")
    Call<Requisition> createRequisition(@Field("institution_id") String institution_id,
                                        @Field("redirect") String redirect,
                                        @Field("agreement") String agreement,
                                        @Field("user_language") String user_language,
                                        @Field("reference") String reference,
                                        @Field("account_selection") Boolean account_selection,
                                        @Field("redirect_immediate") Boolean redirect_immediate);


    @GET("requisitions/{id}")
    Call<Requisition> getRequisitionById(@Path("id") String id);

    @GET("accounts/{id}/transactions")
    Call<TransactionsResponse> getAccountTransactions(@Path("id") String id);

    @GET("accounts/{id}/details")
    Call<AccountDetails> getAccountDetails(@Path("id") String id);

    @GET("accounts/{id}/balances")
    Call<Object> getAccountBalances(@Path("id") String id);

}

File: app/src/main/java/com/rokudo/xpense/data/retrofit/GetDataService.kt
package com.rokudo.xpense.data.retrofit

import com.rokudo.xpense.data.retrofit.models.*
import retrofit2.Call
import retrofit2.http.*

interface GetDataService {

    @GET("/api/v2/institutions/?country=RO")
    fun getAllInstitutions(): Call<List<Institution>>

    @FormUrlEncoded
    @POST("/api/v2/token/new/")
    fun getToken(
        @Field("secret_id") secretId: String,
        @Field("secret_key") secretKey: String
    ): Call<Token>

    @FormUrlEncoded
    @POST("/api/v2/token/refresh/")
    fun refreshToken(@Field("refresh") refresh: String): Call<String>

    @FormUrlEncoded
    @POST("/api/v2/agreements/enduser/")
    fun createEUA(
        @Field("institution_id") institutionId: String,
        @Field("access_scope") accessScope: List<String>
    ): Call<EndUserAgreement>

    @FormUrlEncoded
    @POST("/api/v2/requisitions/")
    fun createRequisition(
        @Field("institution_id") institutionId: String,
        @Field("redirect") redirect: String,
        @Field("agreement") agreement: String,
        @Field("user_language") userLanguage: String,
        @Field("reference") reference: String,
        @Field("account_selection") accountSelection: Boolean?,
        @Field("redirect_immediate") redirectImmediate: Boolean?
    ): Call<Requisition>

    @GET("/api/v2/requisitions/{id}/")
    fun getRequisitionById(@Path("id") id: String): Call<Requisition>

    @GET("/api/v2/requisitions/")
    fun getRequisitions(): Call<RequisitionsResult>

    @DELETE("/api/v2/requisitions/{id}/")
    fun deleteRequisition(@Path("id") id: String): Call<DeleteResponse>

    @DELETE("/api/v2/agreements/enduser/{id}/")
    fun deleteEua(@Path("id") id: String): Call<DeleteResponse>

    @GET("/api/v2/accounts/{id}/transactions/")
    fun getAccountTransactions(
        @Path("id") id: String,
        @Query("date_from") dateFrom: String
    ): Call<TransactionsResponse>

    @GET("/api/v2/accounts/{id}/details/")
    fun getAccountDetails(@Path("id") id: String): Call<AccountDetails>

    @GET("/api/v2/accounts/{id}/balances/")
    fun getAccountBalances(@Path("id") id: String): Call<Balances>
}

